File: web/js/rules.js
// 목표 대비 판정 -> 매수/매도 안내. 예외(BYPASS) 처리 포함.
//
// 목표는 3가지 방식으로 걸 수 있다.
//   weight : 비중 %      (허용오차 tolerance 는 %p)
//   amount : 투자금액     (허용오차는 목표금액의 %)
//   shares : 주 수량      (허용오차는 목표주수의 %, 종목 축에서만 의미 있음)

import { DIM_LABELS, MODE_LABELS } from './util.js';

export const STATUS = { ACTIVE: '실행', BYPASSED: '예외', MUTED: '참고', OK: '적정' };
export const ACTION = { BUY: '매수', SELL: '매도', HOLD: '유지' };

export function dimLabel(dim) { return DIM_LABELS[dim] || dim; }
export function modeLabel(mode) { return MODE_LABELS[mode] || mode; }

// 그룹 기본 허용오차 (설정에 없을 때)
const DEFAULT_TOLERANCE = { weight: 5, amount: 10, shares: 10 };

export function bandOf(item, groupTolerance) {
  const mode = item.mode || 'weight';
  const tol = num(item.tolerance) ?? num(groupTolerance) ?? DEFAULT_TOLERANCE[mode];
  const target = num(item.target);
  let lo = num(item.min);
  let hi = num(item.max);
  if (target !== null) {
    if (lo === null) lo = mode === 'weight' ? Math.max(0, target - tol) : target * (1 - tol / 100);
    if (hi === null) hi = mode === 'weight' ? target + tol : target * (1 + tol / 100);
  }
  return { lo, hi, tol, mode, target };
}

// 축/모드에 따라 "지금 값"을 뽑는다
export function currentValue(bucket, mode) {
  if (!bucket) return 0;
  if (mode === 'amount') return bucket.marketValue || 0;
  if (mode === 'shares') return bucket.quantity || 0;
  return bucket.weight || 0;
}

export function formatValue(v, mode, baseCurrency = 'KRW') {
  if (v === null || v === undefined) return '-';
  if (mode === 'weight') return `${v.toFixed(1)}%`;
  if (mode === 'shares') return `${trimNum(v)}주`;
  return `${Math.round(v).toLocaleString('ko-KR')} ${baseCurrency}`;
}

function trimNum(v) {
  return Number.isInteger(v) ? String(v) : v.toFixed(2).replace(/\.?0+$/, '');
}

// ---------------------------------------------------------------- 예외(BYPASS)
export class BypassRegistry {
  constructor(db, { extra = [], disabled = false, today = null } = {}) {
    this.enabled = (db.bypass?.enabled ?? true) && !disabled;
    this.today = today || new Date().toISOString().slice(0, 10);
    this.entries = [...(db.bypass?.entries || []), ...extra];
    this.hits = new Set();
  }

  isActive(entry) {
    if (!entry.until) return true;
    return this.today <= String(entry.until);
  }

  check(dimension, key) {
    if (!this.enabled) return null;
    for (const e of this.entries) {
      if (!this.isActive(e)) continue;
      const scopeOk = e.scope === 'all' || e.scope === '*' || e.scope === dimension;
      const keyOk = !e.key || e.key === '*'
        || String(e.key).toUpperCase() === String(key).toUpperCase();
      if (scopeOk && keyOk) {
        this.hits.add(`${e.scope}:${e.key}`);
        return e;
      }
    }
    return null;
  }

  get unused() {
    return this.entries.filter((e) => !this.isActive(e) || !this.hits.has(`${e.scope}:${e.key}`));
  }
}

// ---------------------------------------------------------------- 판정
export function evaluate(pf, db, bypass = null) {
  const reg = bypass || new BypassRegistry(db);
  const signals = [];
  const total = pf.totalValue;
  if (!total) return { signals, bypass: reg };

  const minTrade = Number(db.rules?.min_trade_amount) || 0;

  for (const [dim, group] of Object.entries(db.targets || {})) {
    if (group?.enabled === false) continue;
    const buckets = new Map((pf.breakdowns[dim] || []).map((b) => [b.key, b]));
    const seen = new Set();

    for (const [key, rawItem] of Object.entries(group.items || {})) {
      seen.add(key);
      const item = { ...rawItem, mode: rawItem.mode || 'weight' };
      const bucket = buckets.get(key);
      const { lo, hi, mode, target } = bandOf(item, group.tolerance);
      const current = currentValue(bucket, mode);

      let action = 'HOLD';
      let anchor = target ?? current;
      if (hi !== null && current > hi) { action = 'SELL'; anchor = target ?? hi; }
      else if (lo !== null && current < lo) { action = 'BUY'; anchor = target ?? lo; }

      const sig = {
        dimension: dim,
        key,
        label: bucket?.label || key,
        mode,
        action,
        status: action === 'HOLD' ? 'OK' : 'ACTIVE',
        current,
        target,
        min: lo,
        max: hi,
        gap: target === null ? null : current - target,
        amountBase: 0,
        shares: null,
        reason: item.note || '',
        candidates: [],
        candidateTickers: [],
      };

      if (action !== 'HOLD') {
        const diff = Math.abs(current - anchor);
        if (mode === 'weight') sig.amountBase = (diff / 100) * total;
        else if (mode === 'amount') sig.amountBase = diff;
        else sig.amountBase = diff * priceBaseOf(pf, key);
      }
      finalize(sig, pf, db, reg, item.bypass, minTrade);
      signals.push(sig);
    }

    // 목표를 안 정한 그룹도 참고용으로 보여준다
    for (const [key, bucket] of buckets) {
      if (seen.has(key)) continue;
      signals.push({
        dimension: dim, key, label: bucket.label, mode: 'weight', action: 'HOLD',
        status: 'MUTED', current: bucket.weight, target: null, min: null, max: null,
        gap: null, amountBase: 0, shares: null, reason: '목표 미설정',
        candidates: [], candidateTickers: [],
      });
    }
  }

  signals.push(...positionRules(pf, db, reg, minTrade));

  const order = { SELL: 0, BUY: 1, HOLD: 2 };
  const rank = { ACTIVE: 0, BYPASSED: 1, MUTED: 2, OK: 3 };
  signals.sort((a, b) => (rank[a.status] - rank[b.status])
    || (order[a.action] - order[b.action])
    || (b.amountBase - a.amountBase));
  return { signals, bypass: reg };
}

function priceBaseOf(pf, ticker) {
  const p = pf.positions.find((x) => x.ticker === ticker);
  return p && p.hasPrice ? p.priceLocal * p.fx : 0;
}

// 종목 단위 안전장치 (비중 상·하한, 익절, 손절)
function positionRules(pf, db, reg, minTrade) {
  const out = [];
  const r = db.rules || {};
  const maxW = num(r.max_position_weight);
  const minW = num(r.min_position_weight);
  const tp = num(r.take_profit_pct);
  const sl = num(r.stop_loss_pct);

  for (const p of pf.positions) {
    if (!p.hasPrice) continue;
    const name = p.asset.name;
    const push = (action, amount, reason) => {
      const sig = {
        // 개별규칙의 '현재값'은 비중(%)이고, 조정 금액만 기준통화다
        dimension: 'rule', key: p.ticker, label: name, mode: 'weight', action,
        status: 'ACTIVE', current: p.weight, target: null, min: null, max: null,
        gap: null, amountBase: amount, shares: null, reason,
        candidates: [], candidateTickers: [],
      };
      finalize(sig, pf, db, reg, false, minTrade);
      out.push(sig);
    };

    if (maxW !== null && p.weight > maxW) {
      push('SELL', ((p.weight - maxW) / 100) * pf.totalValue, `1종목 비중 상한 ${maxW}% 초과`);
    }
    if (minW !== null && p.weight > 0 && p.weight < minW) {
      push('BUY', ((minW - p.weight) / 100) * pf.totalValue,
        `자투리 종목 - 비중 하한 ${minW}% 미만 (정리 또는 추가매수)`);
    }
    if (p.returnPct === null) continue;
    if (tp !== null && p.returnPct >= tp) {
      push('SELL', (p.marketValueBase || 0) * 0.25,
        `목표 수익률 도달 (${p.returnPct.toFixed(1)}% ≥ ${tp}%) - 일부 익절 검토`);
    }
    if (sl !== null && p.returnPct <= sl) {
      push('SELL', p.marketValueBase || 0,
        `손절선 이탈 (${p.returnPct.toFixed(1)}% ≤ ${sl}%) - 대응 필요`);
    }
  }
  return out;
}

function finalize(sig, pf, db, reg, itemBypass, minTrade) {
  if (sig.action === 'BUY' || sig.action === 'SELL') attachExecution(sig, pf);
  if (sig.status !== 'ACTIVE') return;

  if (itemBypass) {
    sig.status = 'BYPASSED';
    sig.reason = join(sig.reason, '설정에서 이 항목 예외 처리');
    return;
  }
  let hit = reg.check(sig.dimension, sig.key);
  if (!hit && sig.dimension === 'rule') hit = reg.check('ticker', sig.key);
  if (hit) {
    sig.status = 'BYPASSED';
    sig.reason = join(sig.reason, `예외: ${hit.reason || '사유 미기재'}${hit.until ? `, ~${hit.until}` : ''}`);
    return;
  }
  if (minTrade && sig.amountBase < minTrade) {
    sig.status = 'MUTED';
    sig.reason = join(sig.reason, `조정금액이 최소 매매금액(${minTrade.toLocaleString('ko-KR')}) 미만`);
  }
}

function attachExecution(sig, pf) {
  const direct = pf.positions.find((p) => p.ticker === sig.key);
  if (direct) {
    const priceBase = direct.hasPrice ? direct.priceLocal * direct.fx : 0;
    if (priceBase) {
      let shares = sig.amountBase / priceBase;
      if (sig.action === 'SELL') shares = Math.min(shares, direct.quantity);
      sig.shares = shares;
    }
    sig.candidates = [direct.asset.name];
    sig.candidateTickers = [sig.key];
    return;
  }
  // 국가/섹터 같은 그룹 신호는 그룹 안에서 실행 후보를 골라준다
  const members = pf.positions.filter((p) => memberOf(p, sig.dimension, sig.key) && p.hasPrice);
  if (!members.length) return;
  members.sort((a, b) => (sig.action === 'SELL'
    ? (b.returnPct ?? 0) - (a.returnPct ?? 0) || b.weight - a.weight   // 많이 오른 것부터 덜어냄
    : (a.returnPct ?? 0) - (b.returnPct ?? 0) || a.weight - b.weight)); // 덜 오른 것부터 채움
  sig.candidates = members.slice(0, 3).map((p) => p.asset.name);
  sig.candidateTickers = members.slice(0, 3).map((p) => p.ticker);
}

function memberOf(pos, dim, key) {
  switch (dim) {
    case 'country': return pos.asset.country === key;
    case 'sector': return (pos.asset.sector || '기타') === key;
    case 'currency': return pos.asset.currency === key;
    case 'asset_class': return (pos.asset.asset_class || '주식') === key;
    case 'ticker': return pos.ticker === key;
    case 'account': return pos.accounts.includes(key);
    case 'tag': return (pos.asset.tags || []).includes(key);
    default: return false;
  }
}

// ---------------------------------------------------------------- 주문서 합치기
// 국가 초과 + 섹터 초과 + 종목 상한이 사실상 같은 매도를 가리키므로 금액을 더하지 않고
// 가장 큰 제약 하나만 남긴다. 매수/매도가 동시에 걸리면 상계한다.
export function consolidate(signals, pf) {
  const byTicker = new Map();
  for (const s of signals) {
    if (s.status !== 'ACTIVE' || (s.action !== 'BUY' && s.action !== 'SELL')) continue;
    const ticker = s.candidateTickers[0] || s.key;
    if (!byTicker.has(ticker)) byTicker.set(ticker, {});
    const slot = byTicker.get(ticker);
    const tag = `${dimLabel(s.dimension)}:${s.label}${s.reason ? ` (${s.reason.split(' / ')[0]})` : ''}`;
    const cur = slot[s.action] || { amount: 0, reasons: [] };
    slot[s.action] = { amount: Math.max(cur.amount, s.amountBase), reasons: [...cur.reasons, tag] };
  }

  const plan = [];
  for (const [ticker, actions] of byTicker) {
    const sell = actions.SELL || { amount: 0, reasons: [] };
    const buy = actions.BUY || { amount: 0, reasons: [] };
    const netted = Boolean(sell.amount && buy.amount);
    const diff = sell.amount - buy.amount;
    const action = diff > 0 ? 'SELL' : 'BUY';
    const amount = Math.abs(diff);
    if (amount < 1) continue;

    const pos = pf.positions.find((p) => p.ticker === ticker);
    let shares = null;
    if (pos?.hasPrice) {
      const priceBase = pos.priceLocal * pos.fx;
      if (priceBase) {
        shares = amount / priceBase;
        if (action === 'SELL') shares = Math.min(shares, pos.quantity);
      }
    }
    plan.push({
      ticker,
      label: pos ? pos.asset.name : ticker,
      action,
      amountBase: amount,
      shares,
      netted,
      reasons: [...sell.reasons, ...buy.reasons],
    });
  }
  plan.sort((a, b) => (a.action === b.action ? b.amountBase - a.amountBase
    : a.action === 'SELL' ? -1 : 1));
  return plan;
}

function num(v) {
  if (v === null || v === undefined || v === '') return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function join(a, b) { return a ? `${a} / ${b}` : b; }

File: portfolio/rules.py
"""리밸런싱 규칙 엔진.

국가별 / 종목별 / 섹터별 목표 비중을 벗어나면 매수·매도 신호를 만든다.
모든 신호는 bypass(예외 처리) 로 무력화할 수 있고, 무력화된 신호도 리포트에는
'BYPASS + 사유' 로 남겨서 왜 안내가 안 나왔는지 추적 가능하게 한다.

status 값
  ACTIVE   실제로 실행을 권하는 신호
  BYPASSED 사용자가 예외 처리해서 실행 대상에서 제외
  MUTED    조건은 걸렸지만 금액이 너무 작거나(min_trade_amount) 목표 미설정이라 참고용
"""

from __future__ import annotations

import datetime as dt

from .config import BypassEntry, Settings, DIMENSION_LABELS
from .engine import Portfolio
from .models import PlanItem, Signal

DEFAULT_MIN_TRADE = 0.0


class BypassRegistry:
    """설정 파일 + CLI 옵션에서 모은 예외 규칙."""

    def __init__(self, settings: Settings, extra: list[str] | None = None,
                 disabled: bool = False, today: dt.date | None = None):
        self.enabled = settings.bypass_enabled and not disabled
        self.today = today or dt.date.today()
        self.entries: list[BypassEntry] = list(settings.bypass_entries)
        for spec in extra or []:
            self.entries.append(_parse_bypass_spec(spec))
        self.hits: set[tuple[str, str]] = set()

    def check(self, dimension: str, key: str) -> BypassEntry | None:
        if not self.enabled:
            return None
        for e in self.entries:
            if not e.active(self.today):
                continue
            scope_ok = e.scope in ("all", "*") or e.scope == dimension
            key_ok = e.key in ("*", "") or e.key.upper() == key.upper()
            if scope_ok and key_ok:
                self.hits.add((e.scope, e.key))
                return e
        return None

    @property
    def unused(self) -> list[BypassEntry]:
        """설정했지만 이번에 한 번도 안 걸린 예외 (오타/기간만료 점검용)."""
        return [
            e for e in self.entries
            if not e.active(self.today) or (e.scope, e.key) not in self.hits
        ]


def _parse_bypass_spec(spec: str) -> BypassEntry:
    """'TSLA' / 'ticker:TSLA' / 'country:VN=환전 이슈' 형태를 파싱."""
    scope, _, rest = spec.partition(":")
    if not rest:
        scope, rest = "ticker", scope
    key, _, reason = rest.partition("=")
    return BypassEntry(scope=scope.strip(), key=key.strip(),
                       reason=reason.strip() or "CLI 옵션으로 예외 처리")


def evaluate(pf: Portfolio, settings: Settings,
             bypass: BypassRegistry | None = None) -> list[Signal]:
    bypass = bypass or BypassRegistry(settings)
    signals: list[Signal] = []
    total = pf.total_value
    if not total:
        return signals

    min_trade = float(settings.rules.get("min_trade_amount", DEFAULT_MIN_TRADE) or 0)

    for dim, group in settings.targets.items():
        if not group.enabled:
            continue
        buckets = {b.key: b for b in pf.breakdown(dim)}
        seen = set()

        for key, item in group.items.items():
            seen.add(key)
            bucket = buckets.get(key)
            current = bucket.weight if bucket else 0.0
            lo, hi = item.band(group.tolerance)
            label = bucket.label if bucket else key

            if hi is not None and current > hi:
                action, anchor = "SELL", (item.target if item.target is not None else hi)
            elif lo is not None and current < lo:
                action, anchor = "BUY", (item.target if item.target is not None else lo)
            else:
                action, anchor = "HOLD", (item.target if item.target is not None else current)

            gap_pp = current - (item.target if item.target is not None else current)
            amount = abs(current - anchor) / 100.0 * total if action != "HOLD" else 0.0

            sig = Signal(
                dimension=dim,
                key=key,
                label=label,
                action=action,
                status="ACTIVE" if action != "HOLD" else "OK",
                current_weight=current,
                target_weight=item.target,
                min_weight=lo,
                max_weight=hi,
                gap_pp=gap_pp,
                amount_base=amount,
                reason=item.note,
            )
            _finalize(sig, pf, settings, bypass, item.bypass, min_trade)
            signals.append(sig)

        # 목표를 안 정한 그룹도 화면에는 보여준다 (참고용)
        for key, bucket in buckets.items():
            if key in seen:
                continue
            signals.append(Signal(
                dimension=dim, key=key, label=bucket.label, action="HOLD", status="MUTED",
                current_weight=bucket.weight, reason="목표 비중 미설정",
            ))

    signals.extend(_position_rules(pf, settings, bypass, min_trade))
    order = {"SELL": 0, "BUY": 1, "HOLD": 2}
    signals.sort(key=lambda s: (
        0 if s.status == "ACTIVE" else 1 if s.status == "BYPASSED" else 2,
        order.get(s.action, 3),
        -(s.amount_base or 0),
    ))
    return signals


def _position_rules(pf: Portfolio, settings: Settings, bypass: BypassRegistry,
                    min_trade: float) -> list[Signal]:
    """비중 상·하한, 익절/손절 같은 종목 단위 안전장치."""
    out: list[Signal] = []
    r = settings.rules
    max_w = _f(r.get("max_position_weight"))
    min_w = _f(r.get("min_position_weight"))
    tp = _f(r.get("take_profit_pct"))
    sl = _f(r.get("stop_loss_pct"))
    total = pf.total_value

    for p in pf.positions:
        t = p.asset.ticker
        name = f"{p.asset.name}({t})"

        if max_w is not None and p.weight > max_w:
            amount = (p.weight - max_w) / 100.0 * total
            s = Signal("rule", t, name, "SELL", "ACTIVE",
                       current_weight=p.weight, max_weight=max_w, amount_base=amount,
                       reason=f"1종목 비중 상한 {max_w:g}% 초과")
            _finalize(s, pf, settings, bypass, False, min_trade)
            out.append(s)

        if min_w is not None and 0 < p.weight < min_w:
            amount = (min_w - p.weight) / 100.0 * total
            s = Signal("rule", t, name, "BUY", "ACTIVE",
                       current_weight=p.weight, min_weight=min_w, amount_base=amount,
                       reason=f"자투리 종목 - 비중 하한 {min_w:g}% 미만 (정리 또는 추가매수)")
            _finalize(s, pf, settings, bypass, False, min_trade)
            out.append(s)

        ret = p.return_pct
        if ret is None:
            continue
        if tp is not None and ret >= tp:
            s = Signal("rule", t, name, "SELL", "ACTIVE",
                       current_weight=p.weight, amount_base=(p.market_value_base or 0) * 0.25,
                       reason=f"목표 수익률 도달 (+{ret:.1f}% ≥ {tp:g}%) - 일부 익절 검토")
            _finalize(s, pf, settings, bypass, False, min_trade)
            out.append(s)
        if sl is not None and ret <= sl:
            s = Signal("rule", t, name, "SELL", "ACTIVE",
                       current_weight=p.weight, amount_base=(p.market_value_base or 0),
                       reason=f"손절선 이탈 ({ret:.1f}% ≤ {sl:g}%) - 대응 필요")
            _finalize(s, pf, settings, bypass, False, min_trade)
            out.append(s)
    return out


def _finalize(sig: Signal, pf: Portfolio, settings: Settings,
              bypass: BypassRegistry, item_bypass: bool, min_trade: float) -> None:
    """주수 환산 + 후보 종목 + bypass 적용."""
    if sig.action in ("BUY", "SELL"):
        _attach_execution(sig, pf)

    if sig.status != "ACTIVE":
        return

    if item_bypass:
        sig.status = "BYPASSED"
        sig.reason = _join(sig.reason, "설정에서 이 항목 예외 처리(bypass: true)")
        return

    hit = bypass.check(sig.dimension, sig.key)
    if hit is None and sig.dimension == "rule":
        hit = bypass.check("ticker", sig.key)
    if hit:
        sig.status = "BYPASSED"
        until = f", ~{hit.until}" if hit.until else ""
        sig.reason = _join(sig.reason, f"BYPASS: {hit.reason or '사유 미기재'}{until}")
        return

    if min_trade and sig.amount_base < min_trade:
        sig.status = "MUTED"
        sig.reason = _join(sig.reason, f"조정금액이 최소 매매금액({min_trade:,.0f}) 미만")


def _attach_execution(sig: Signal, pf: Portfolio) -> None:
    """신호를 '몇 주' 수준까지 내려준다."""
    by_ticker = {p.asset.ticker: p for p in pf.positions}
    pos = by_ticker.get(sig.key)

    if pos is not None:
        price_base = (pos.price_local or 0) * pos.fx_rate
        if price_base:
            shares = sig.amount_base / price_base
            if sig.action == "SELL":
                shares = min(shares, pos.quantity)
            sig.shares = shares
        sig.candidates = [sig.key]
        sig.candidate_tickers = [sig.key]
        return

    # 국가/섹터 같은 그룹 신호: 그룹 안에서 실행 후보를 골라준다.
    members = [p for p in pf.positions if _member_of(p, sig.dimension, sig.key)]
    if not members:
        return
    if sig.action == "SELL":
        # 많이 오른 / 비중 큰 종목부터 덜어내는 게 자연스럽다
        members.sort(key=lambda p: (p.return_pct or 0, p.weight), reverse=True)
    else:
        # 덜 오른 / 비중 작은 종목부터 채운다
        members.sort(key=lambda p: (p.return_pct if p.return_pct is not None else 0, p.weight))
    sig.candidates = [f"{p.asset.name}({p.asset.ticker})" for p in members[:3]]
    sig.candidate_tickers = [p.asset.ticker for p in members[:3]]


def _member_of(pos, dimension: str, key: str) -> bool:
    getters = {
        "country": lambda p: p.asset.country,
        "sector": lambda p: p.asset.sector,
        "currency": lambda p: p.asset.currency,
        "asset_class": lambda p: p.asset.asset_class,
        "ticker": lambda p: p.asset.ticker,
    }
    if dimension == "tag":
        return key in pos.asset.tags
    if dimension == "account":
        return key in pos.accounts
    fn = getters.get(dimension)
    return bool(fn and fn(pos) == key)


def action_plan(signals: list[Signal]) -> list[Signal]:
    """실제로 실행을 권하는 신호만 (원본 그대로)."""
    return [s for s in signals if s.is_actionable]


def consolidate(signals: list[Signal], pf: Portfolio) -> list[PlanItem]:
    """겹치는 규칙을 종목 단위 주문서 한 장으로 합친다.

    - 같은 종목에 같은 방향 신호가 여러 개면 '가장 큰 금액'을 쓴다(합산 아님).
      국가 초과 + 섹터 초과 + 종목 초과가 사실상 같은 매도를 가리키기 때문.
    - 매수/매도가 동시에 걸리면 상계해서 순매매만 남긴다.
    """
    by_ticker: dict[str, dict[str, tuple[float, list[str]]]] = {}

    for s in signals:
        if not s.is_actionable:
            continue
        ticker = s.candidate_tickers[0] if s.candidate_tickers else s.key
        tag = f"{dimension_label(s.dimension)}:{s.label}"
        if s.reason:
            tag += f" ({s.reason.split(' / ')[0]})"
        slot = by_ticker.setdefault(ticker, {})
        amount, reasons = slot.get(s.action, (0.0, []))
        slot[s.action] = (max(amount, s.amount_base), reasons + [tag])

    positions = {p.asset.ticker: p for p in pf.positions}
    plan: list[PlanItem] = []
    for ticker, actions in by_ticker.items():
        sell_amt, sell_reasons = actions.get("SELL", (0.0, []))
        buy_amt, buy_reasons = actions.get("BUY", (0.0, []))
        netted = bool(sell_amt and buy_amt)
        amount = sell_amt - buy_amt
        action = "SELL" if amount > 0 else "BUY"
        amount = abs(amount)
        if amount < 1e-6:
            continue

        pos = positions.get(ticker)
        label = f"{pos.asset.name}({ticker})" if pos else ticker
        shares = None
        if pos:
            price_base = (pos.price_local or 0) * pos.fx_rate
            if price_base:
                shares = amount / price_base
                if action == "SELL":
                    shares = min(shares, pos.quantity)
        plan.append(PlanItem(
            ticker=ticker, label=label, action=action, amount_base=amount,
            shares=shares, reasons=(sell_reasons + buy_reasons), netted=netted,
        ))

    plan.sort(key=lambda i: (0 if i.action == "SELL" else 1, -i.amount_base))
    return plan


def dimension_label(dim: str) -> str:
    return DIMENSION_LABELS.get(dim, "개별규칙" if dim == "rule" else dim)


def _f(v):
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def _join(a: str, b: str) -> str:
    return f"{a} / {b}" if a else b
